fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+		 */
+
+		ret = test->rx_transfer(arguments->test_arguments,
+				state->cur_sender, 1,
+				domain_state->addresses[state->cur_sender],
+				domain_state->endpoint,
+				op_context, op_context->buf,
+				NULL, state->rx_flags);
+
+		if (ret == -FI_EAGAIN) {
+			state->cur_sender = prev;
+			state->cur_sender_rx_threshold = prev_threshold;
+			break;
+		}
+
+		hpcs_verbose("rx_transfer initiated: ctx %p "
+			     "from rank %ld\n",
+			     op_context, state->cur_sender);
+
+		if (ret) {
+			hpcs_error("test receive failed, ret=%d\n", ret);
+			return ret;
+		}
+
+		op_context->state = OP_PENDING;
+		op_context->core_context = state->recvs_posted;
+
+		state->recvs_posted++;
+		state->rx_window--;
+	};
+
+	return 0;
+}
+
+/*
+ * Initiate as many tx transfers as our window allows, within
+ * a single iteration of test/pattern.
+ *
+ * Return 0 unless an error occurs.
+ */
+
+static int core_initiate_tx(struct domain_state *domain_state,
+		struct arguments *arguments,
+		struct pattern_api *pattern,
+		struct test_api *test,
+		struct test_config *test_config,
+		struct job *job,
+		struct core_state *state)
+{
+	int ret, i, prev, prev_threshold;
+	struct op_context* op_context;
+	enum callback_order order = arguments->callback_order;
+
+	struct fid_mr* mr;
+	void *mr_desc;
+
+	/* post send(s) */
+	while (!state->all_sends_done) {
+		if (order == CALLBACK_ORDER_EXPECTED && !state->all_recvs_done)
+			break;
+
+		prev = state->cur_receiver;
+		prev_threshold = state->cur_receiver_tx_threshold;
+
+		ret = pattern->next_receiver(arguments->pattern_arguments,
+				job->rank, job->ranks, &state->cur_receiver,
+				&state->cur_receiver_tx_threshold);
+		if (ret == -ENODATA) {
+			if (order == CALLBACK_ORDER_UNEXPECTED)
+				job->barrier();
+			state->all_sends_done = true;
+			break;
+		} else if (ret < 0) {
+			hpcs_error("next_receiver failed\n");
+			return ret;
+		}
+
+		if (state->tx_window == 0) {
+			state->cur_receiver = prev;
+			state->cur_receiver_tx_threshold = prev_threshold;
+			break;
+		}
+
+		op_context = CONTEXT(state->tx_context, state->sends_posted);
+		if (op_context->state != OP_DONE) {
+			state->cur_receiver = prev;
+			state->cur_receiver_tx_threshold = prev_threshold;
+			break;
+		}
+
+		test->tx_init_buffer(arguments->test_arguments,
+				DATA_BUF(state->tx_buf, state->sends_posted),
+				test_config->tx_buffer_size);
+
+		if (test_config->tx_use_cntr) {
+			mr = test->tx_create_mr(arguments->test_arguments,
+					domain_state->domain, 0,
+					DATA_BUF(state->tx_buf, state->sends_posted),
+					arguments->buffer_size, FI_SEND, 0);
+			if (mr == NULL) {
+				ret = -1;
+				hpcs_error("unable to register tx memory region\n");
+				return ret;
+			}
+
+			mr_desc = fi_mr_desc(mr);
+		} else {
+			mr = NULL;
+			mr_desc = NULL;
+		}
+
+		op_context->buf = DATA_BUF(state->tx_buf, state->sends_posted);
+		op_context->tx_mr = mr;
+
+		if (pattern->enable_triggered) {
```

Comment:
I'd personally like to see triggered support dropped completely from the API, so I would not worry about testing it.  Few providers support it, and maintaining it as part of testing is difficult as a result.